lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp

```diff
@@ -856,199 +870,238 @@ void mlir::triton::updateWaits(ModuleOp module) {
-  int argIdx = -1;
-  auto iter = std::find(yieldOp->getOperands().begin(),
-                        yieldOp->getOperands().end(), result);
-  if (iter != yieldOp->getOperands().end())
-    argIdx = std::distance(yieldOp->getOperands().begin(), iter);
-  if (argIdx == -1)
-    return false;
-  for (auto operand : dotOp.getOperands()) {
-    if (dependOn(operand, argIdx, forOp)) {
-      auto iterArgs = forOp.getRegionIterArgs();
-      *firstUse = iterArgs[argIdx].use_begin().getUser();
-      return true;
+// Tries to cast op to a tt::DotOp, if it's an MMAv3 (i.e. wgmma / hopper) dot.
+//
+// Note that this returns nullptr for ttng::DotAsyncOps, but we can't call this
+// "dynCastMMAv3SyncDot", because this is sometimes used before we've converted
+// eligible tt::DotOps to ttng::DotAsyncOp.
+tt::DotOp dynCastMMAv3Dot(Operation *op) {
+  if (auto dotOp = dyn_cast<tt::DotOp>(op)) {
+    auto resEnc =
+        dotOp.getType().getEncoding().dyn_cast<ttg::NvidiaMmaEncodingAttr>();
+    if (resEnc && resEnc.isHopper()) {
+      return dotOp;
     }
   }
-  return false;
+  return nullptr;
 }
 
-static void removeExtraWait(tt::nvidia_gpu::DotWaitOp dotWaitOp,
-                            bool hasDotWait0) {
-  if (hasDotWait0) {
-    dotWaitOp->erase();
+// Determines whether a tt.dot op can be transformed into ttng.dot.async.  The
+// dot can be made async if all of the following are true.
+//
+//  0. The operation is an MMAv3 dot.
+//
+//  1. All operands that touch shared memory are used "safely", i.e. can't read
+//     an incomplete value while it's being written asynchronously by a load.
+//
+//  2. During iteration i, nothing other than the loop's `yield` reads the
+//     result of the dot.
+//
+//  3. During iteration i, the result of the dot from iteration i-1 is consumed
+//     only by other MMAv3 dots (either sync or async) as the `c` operand.
+//
+//     This is safe because the following pseudo-PTX is valid (even though the
+//     docs don't say so explicitly):
+//
+//        %accum = dot.async %a1, %b1, %c1
+//        %accum = dot.async %a2, %b2, %accum
+//
+//     That is, the second async dot can use the result of the first one without
+//     an intervening wait.  However, the only operation that can legally read
+//     %accum before the wait is another dot.async, and our current
+//     understanding is this only works for the `c` operand, not `a` or `b`.
+//
+//     A synchronous MMAv3 dot will be codegen'ed as `dot.async; wait 0`, so it
+//     works just as well as an async dot for these purposes.
+//
+// On success, this function returns the index of the dot in the loop's
+// iter_args.  (Rule (2) above ensures this is well-defined.)
+static std::optional<int> dotCanBeMadeAsync(Operation *op, scf::ForOp forOp) {
+  // Rule 0: The operation is an MMAv3 dot.
+  tt::DotOp dotOp = dynCastMMAv3Dot(op);
+  if (!dotOp)
+    return std::nullopt;
+
+  // Rule 1: All shmem operands are used "safely".
```

Comment:
nit: I would replace used "safely" by "have been multi-buffered"

Reply:
not sure if you missed that one or prefer to leave it as is, commenting just in case

Reply:
Thanks, fixed.